pkg/api/customization/cluster/actions_rketemplate.go

```diff
@@ -0,0 +1,159 @@
+package cluster
+
+import (
+	"fmt"
+	"net/http"
+
+	"time"
+
+	"github.com/rancher/norman/api/access"
+	"github.com/rancher/norman/httperror"
+	"github.com/rancher/norman/parse"
+	"github.com/rancher/norman/types"
+	"github.com/rancher/rancher/pkg/controllers/management/globalnamespacerbac"
+	"github.com/rancher/rancher/pkg/namespace"
+	"github.com/rancher/rancher/pkg/ref"
+	v3 "github.com/rancher/types/apis/management.cattle.io/v3"
+	mgmtclient "github.com/rancher/types/client/management/v3"
+	"github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func (a ActionHandler) saveAsTemplate(actionName string, action *types.Action, apiContext *types.APIContext) error {
+
+	var clusterForAccessCheck mgmtclient.Cluster
+	var err error
+	if err := access.ByID(apiContext, apiContext.Version, apiContext.Type, apiContext.ID, &clusterForAccessCheck); err != nil {
+		return httperror.NewAPIError(httperror.NotFound,
+			fmt.Sprintf("failed to get cluster by id %v", apiContext.ID))
+	}
+
+	cluster, err := a.ClusterClient.Get(apiContext.ID, v1.GetOptions{})
+	if err != nil {
+		return httperror.WrapAPIError(err, httperror.NotFound,
+			fmt.Sprintf("cluster with id %v doesn't exist", apiContext.ID))
+	}
+
+	if cluster.DeletionTimestamp != nil {
+		return httperror.NewAPIError(httperror.InvalidType,
+			fmt.Sprintf("cluster with id %v is being deleted", apiContext.ID))
+	}
+	if !v3.ClusterConditionReady.IsTrue(cluster) {
+		return httperror.WrapAPIError(err, httperror.ClusterUnavailable,
+			fmt.Sprintf("cluster not ready"))
+	}
+
+	//Copy the cluster rke spec to a template and revision object
+
+	actionInput, err := parse.ReadBody(apiContext.Request)
+	if err != nil {
+		return err
+	}
+	clusterTempName, ok := actionInput["clusterTemplateName"].(string)
```

Comment:
do you think it would make sense to validate input params before trying to fetch the cluster from api? So in case input is invalid, we can save on api call

Reply:
Also would be nice to split this method to multiple subcalls, as it's a bit long. I can think of validateActionInput, validateCluster, createClusterTemplateRevision, updateCluster.

Reply:
done